src/tasks/migration/DiagnosticProvider.ts

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import * as vscode from "vscode";
+import { asyncDebounce, getTasksJsonPaths } from "./utils";
+import * as readline from "readline";
+import * as fs from "fs";
+import { DeprecatedExportJarTaskProvider, ExportJarTaskProvider } from "../../exportJarSteps/ExportJarTaskProvider";
+
+export class DiagnosticProvider {
+    public static DIAGNOSTIC_SOURCE = "Project Manager for Java";
+    public static DEPRECATED_TASK_TYPE_DECLARATION = `\"type\": \"${DeprecatedExportJarTaskProvider.type}\"`;
+    public static DEPRECATED_TASK_TYPE_MESSAGE = `Tasks with type \"${DeprecatedExportJarTaskProvider.type}\" are deprecated and will not be supported in the future, please use \"${ExportJarTaskProvider.exportJarType}\" instead.`;
+    private diagnosticCollection: vscode.DiagnosticCollection;
+    private refreshDiagnosticsTrigger: any;
+
+    public getDiagnosticCollection() {
+        return this.diagnosticCollection;
+    }
+
+    public async initialize(context: vscode.ExtensionContext): Promise<void> {
+        this.refreshDiagnosticsTrigger = asyncDebounce(
+            this.refreshDiagnostics,
+            500 /** ms */,
+            this
+        );
+        this.diagnosticCollection = vscode.languages.createDiagnosticCollection("migrateExportTask");
+        context.subscriptions.push(this.diagnosticCollection);
+        context.subscriptions.push(vscode.workspace.onDidChangeTextDocument(async (e) => {
+            if (e.document.fileName.includes("tasks.json")) {
+                await this.refreshDiagnosticsTrigger(e.document.uri);
+            }
+        }));
+        await this.initializeDiagnostics();
+    }
+
+    private async initializeDiagnostics(): Promise<void> {
+        const tasksJsonPaths = await getTasksJsonPaths();
+        for (const tasksJsonPath of tasksJsonPaths) {
+            const diagnostics: vscode.Diagnostic[] = await DiagnosticProvider.getDiagnosticsFromTasksJsonPath(tasksJsonPath);
+            this.diagnosticCollection.set(vscode.Uri.file(tasksJsonPath), diagnostics);
+        }
+    }
+
+    private async refreshDiagnostics(uri: vscode.Uri): Promise<void> {
+        const diagnostics: vscode.Diagnostic[] = await DiagnosticProvider.getDiagnosticsFromTasksJsonPath(uri.fsPath);
+        this.diagnosticCollection.set(uri, diagnostics);
+    }
+
+    private static async getDiagnosticsFromTasksJsonPath(tasksJsonPath: string): Promise<vscode.Diagnostic[]> {
+        const diagnostics: vscode.Diagnostic[] = [];
+        const fileStream = fs.createReadStream(tasksJsonPath);
+        let lineNumber = 0;
+        const rl = readline.createInterface({
+            input: fileStream,
+            crlfDelay: Infinity
+        });
+        for await (const line of rl) {
+            const columnNumber = line.indexOf(DiagnosticProvider.DEPRECATED_TASK_TYPE_DECLARATION);
```

Comment:
What if there is no space between `": "`?

Looks like a safer way is to parse the json.

Reply:
I didn't find a way to get the exact row and column when parsing the json, so here just read the file line by line.

Reply:
What about regex match